File: rewrite-javascript/rewrite/test/preconditions.test.ts
/*
 * Copyright 2025 the original author or authors.
 * <p>
 * Licensed under the Moderne Source Available License (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * https://docs.moderne.io/licensing/moderne-source-available-license
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import {RecipeSpec} from "../src/test";
import {javascript} from "../src/javascript";
import {ConditionalFindIdentifier, FindIdentifierWithPathPrecondition} from "../fixtures/path-precondition";
import {and, check, CompositePrecondition, not, or, RecipeRef} from "../src/preconditions";
import {findMethods, findTypes, hasSourcePath, usesMethod, usesType} from "../src/javascript/preconditions";
import {ExecutionContext} from "../src/execution";
import {TreeVisitor} from "../src/visitor";
import {Cursor, Tree} from "../src/tree";

describe('Preconditions', () => {
    test('visitor precondition - should only mark identifiers in matching path', async () => {
        const spec = new RecipeSpec();
        spec.recipe = new FindIdentifierWithPathPrecondition({
            requiredPath: 'test.js',
            identifier: 'foo'
        });

        await spec.rewriteRun(
            {
                ...javascript('const foo = 1;', 'const /*~~>*/foo = 1;'),
                path: 'test.js'
            },
            {
                ...javascript('const foo = 2;'),
                path: 'other.js'
            }
        );
    });

    test('boolean precondition - should mark when true', async () => {
        const spec = new RecipeSpec();
        spec.recipe = new ConditionalFindIdentifier({
            shouldSearch: true,
            identifier: 'bar'
        });

        await spec.rewriteRun(
            javascript('const bar = 1;', 'const /*~~>*/bar = 1;')
        );
    });

    test('boolean precondition - should not mark when false', async () => {
        const spec = new RecipeSpec();
        spec.recipe = new ConditionalFindIdentifier({
            shouldSearch: false,
            identifier: 'bar'
        });

        await spec.rewriteRun(
            javascript('const bar = 1;')
        );
    });
});

/**
 * Stub source-file sentinel: only ``isSourceFile(tree)`` needs to be true
 * inside ``Check.visit``; the wrapper doesn't read source-file fields.
 */
const stubSourceFile = (): Tree => ({
    kind: "org.openrewrite.tree.SourceFile",
    id: "00000000-0000-0000-0000-000000000000",
    markers: {kind: "org.openrewrite.marker.Markers", id: "00000000-0000-0000-0000-000000000000", markers: []},
    sourcePath: "stub",
    charsetName: undefined,
    charsetBomMarked: false,
    checksum: undefined,
    fileAttributes: undefined
} as unknown as Tree);

class RecordingVisitor extends TreeVisitor<any, ExecutionContext> {
    calls = 0;

    async visit<R extends any>(tree: Tree, _: ExecutionContext, _parent?: Cursor): Promise<R | undefined> {
        this.calls++;
        return tree as R;
    }
}

class MarkingVisitor extends TreeVisitor<any, ExecutionContext> {
    calls = 0;

    async visit<R extends any>(tree: Tree, _: ExecutionContext, _parent?: Cursor): Promise<R | undefined> {
        this.calls++;
        return ({...(tree as any)}) as R;
    }
}

describe('Preconditions composites (in-process)', () => {
    test('or short-circuits on the first matching operand', async () => {
        const matching = new MarkingVisitor();
        const nonMatching = new RecordingVisitor();
        const editor = new RecordingVisitor();

        const wrapped = await check(or(matching, nonMatching), editor);
        await wrapped.visit(stubSourceFile(), {} as ExecutionContext);

        expect(matching.calls).toBe(1);
        expect(nonMatching.calls).toBe(0);
        expect(editor.calls).toBe(1);
    });

    test('or skips editor when no operand matches', async () => {
        const a = new RecordingVisitor();
        const b = new RecordingVisitor();
        const editor = new RecordingVisitor();

        const wrapped = await check(or(a, b), editor);
        await wrapped.visit(stubSourceFile(), {} as ExecutionContext);

        expect(a.calls).toBe(1);
        expect(b.calls).toBe(1);
        expect(editor.calls).toBe(0);
    });

    test('and runs editor only when all operands match', async () => {
        const matchA = new MarkingVisitor();
        const matchB = new MarkingVisitor();
        const editor = new RecordingVisitor();

        const wrapped = await check(and(matchA, matchB), editor);
        await wrapped.visit(stubSourceFile(), {} as ExecutionContext);
        expect(editor.calls).toBe(1);

        const editor2 = new RecordingVisitor();
        const wrapped2 = await check(and(matchA, new RecordingVisitor()), editor2);
        await wrapped2.visit(stubSourceFile(), {} as ExecutionContext);
        expect(editor2.calls).toBe(0);
    });

    test('not inverts a single-operand match', async () => {
        const matching = new MarkingVisitor();
        const editor1 = new RecordingVisitor();
        await (await check(not(matching), editor1)).visit(stubSourceFile(), {} as ExecutionContext);
        expect(editor1.calls).toBe(0);

        const nonMatching = new RecordingVisitor();
        const editor2 = new RecordingVisitor();
        await (await check(not(nonMatching), editor2)).visit(stubSourceFile(), {} as ExecutionContext);
        expect(editor2.calls).toBe(1);
    });

    test('or/and require at least two operands', () => {
        expect(() => or(new RecordingVisitor())).toThrow(/at least two operands/);
        expect(() => and(new RecordingVisitor())).toThrow(/at least two operands/);
    });

    test('CompositePrecondition is a plain class', () => {
        const composite = or(new RecordingVisitor(), new MarkingVisitor());
        expect(composite).toBeInstanceOf(CompositePrecondition);
        expect(composite.op).toBe("or");
        expect(composite.operands).toHaveLength(2);
    });

    test('RecipeRef helpers are lazy and identifiable', () => {
        // Helpers must NOT fire any RPC when called — they just bundle a
        // recipe name + options for later wire emission.
        const refs = [
            hasSourcePath("**/foo.ts"),
            usesMethod("*..* tostring(..)"),
            usesType("java.util.List"),
            findMethods("*..* fromstring(..)"),
            findTypes("java.util.Map"),
        ];

        for (const ref of refs) {
            expect(ref).toBeInstanceOf(RecipeRef);
        }

        expect(refs[0].recipeName).toBe("org.openrewrite.FindSourceFiles");
        expect(refs[0].options).toEqual({filePattern: "**/foo.ts"});

        expect(refs[1].recipeName).toBe("org.openrewrite.java.search.HasMethod");
        expect(refs[1].options).toEqual({methodPattern: "*..* tostring(..)", matchOverrides: false});

        expect(refs[2].recipeName).toBe("org.openrewrite.java.search.HasType");
        expect(refs[2].options).toEqual({fullyQualifiedTypeName: "java.util.List", checkAssignability: false});

        expect(refs[3].recipeName).toBe("org.openrewrite.java.search.FindMethods");
        expect(refs[4].recipeName).toBe("org.openrewrite.java.search.FindTypes");
    });

    test('RecipeRef without localVisitor short-circuits to "matches"', async () => {
        // Construct a RecipeRef directly without bundling a native
        // visitor — in-process callers should fall back to "always
        // matches" so the wrapped editor still runs (the host evaluates
        // the gate for real once the response goes over the wire).
        const editor = new RecordingVisitor();
        const refWithoutLocal = new RecipeRef("org.openrewrite.java.search.HasMethod", {
            methodPattern: "*..* nope(..)",
            matchOverrides: false,
        });
        await (await check(refWithoutLocal, editor))
            .visit(stubSourceFile(), {} as ExecutionContext);
        expect(editor.calls).toBe(1);
    });

    test('RecipeRef with localVisitor evaluates the gate for real', async () => {
        // Helpers like usesMethod populate a native local visitor so
        // unit tests still see real filtering behavior. The stub source
        // file has no method invocations, so the gate fails and the
        // editor is skipped.
        const editor = new RecordingVisitor();
        await (await check(usesMethod("*..* nope(..)"), editor))
            .visit(stubSourceFile(), {} as ExecutionContext);
        expect(editor.calls).toBe(0);
    });

    test('helpers populate localVisitor for real in-process filtering', () => {
        // Spot-check that helpers bundle a TreeVisitor for offline
        // eval. The actual filtering behavior is exercised by the
        // RecipeSpec / fixture tests at the top of this file.
        expect(hasSourcePath("**/*.ts").localVisitor).toBeDefined();
        expect(usesMethod("*..* a(..)").localVisitor).toBeDefined();
        expect(usesType("X").localVisitor).toBeDefined();
        expect(findMethods("*..* a(..)").localVisitor).toBeDefined();
        expect(findTypes("X").localVisitor).toBeDefined();
    });
});

File: rewrite-go/pkg/preconditions/preconditions_test.go
/*
 * Copyright 2026 the original author or authors.
 *
 * Licensed under the Moderne Source Available License (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://docs.moderne.io/licensing/moderne-source-available-license
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package preconditions

import (
	"testing"

	"github.com/stretchr/testify/assert"

	"github.com/openrewrite/rewrite/rewrite-go/pkg/tree/golang"
	"github.com/openrewrite/rewrite/rewrite-go/pkg/tree/java"
)

// newSourceFile returns a fresh CompilationUnit pointer to use as the
// source-file sentinel. Each call returns a distinct identity so a
// "marking" visitor can return a different value.
func newSourceFile() java.Tree {
	return &golang.CompilationUnit{}
}

// recordingVisitor is a TreeVisitor that returns the same tree, recording
// each call. Simulates a precondition that does NOT match.
type recordingVisitor struct{ calls int }

func (r *recordingVisitor) Visit(t java.Tree, _ any) java.Tree {
	r.calls++
	return t
}

// markingVisitor is a TreeVisitor that returns a *different* tree
// (any non-identity value), simulating a precondition that DOES match
// (e.g. by adding a SearchResult marker).
type markingVisitor struct{ calls int }

func (m *markingVisitor) Visit(t java.Tree, _ any) java.Tree {
	m.calls++
	return newSourceFile() // different identity from the input
}

func TestCheckRunsEditorWhenConditionMarks(t *testing.T) {
	cond := &markingVisitor{}
	editor := &recordingVisitor{}

	wrapped := Check(cond, editor)
	wrapped.Visit(newSourceFile(), nil)

	assert.Equal(t, 1, cond.calls, "condition calls")
	assert.Equal(t, 1, editor.calls, "editor calls")
}

func TestCheckSkipsEditorWhenConditionReturnsIdentity(t *testing.T) {
	cond := &recordingVisitor{}
	editor := &recordingVisitor{}

	wrapped := Check(cond, editor)
	wrapped.Visit(newSourceFile(), nil)

	assert.Equal(t, 1, cond.calls, "condition calls")
	assert.Equalf(t, 0, editor.calls, "editor calls = %d, want 0 (gate did not match)", editor.calls)
}

func TestOrShortCircuitsOnFirstMatch(t *testing.T) {
	matching := &markingVisitor{}
	nonMatching := &recordingVisitor{}
	editor := &recordingVisitor{}

	Check(Or(matching, nonMatching), editor).Visit(newSourceFile(), nil)

	assert.Equal(t, 1, matching.calls, "matching calls")
	assert.Equalf(t, 0, nonMatching.calls, "nonMatching calls = %d, want 0 (Or should short-circuit)", nonMatching.calls)
	assert.Equal(t, 1, editor.calls, "editor calls")
}

func TestOrSkipsEditorWhenNoOperandMatches(t *testing.T) {
	a := &recordingVisitor{}
	b := &recordingVisitor{}
	editor := &recordingVisitor{}

	Check(Or(a, b), editor).Visit(newSourceFile(), nil)

	assert.False(t, a.calls != 1 || b.calls != 1, "operand calls")
	assert.Equal(t, 0, editor.calls, "editor calls")
}

func TestAndRunsEditorOnlyWhenAllMatch(t *testing.T) {
	a := &markingVisitor{}
	b := &markingVisitor{}
	editor := &recordingVisitor{}
	Check(And(a, b), editor).Visit(newSourceFile(), nil)
	assert.Equal(t, 1, editor.calls, "editor calls (all match")

	editor2 := &recordingVisitor{}
	Check(And(a, &recordingVisitor{}), editor2).Visit(newSourceFile(), nil)
	assert.Equal(t, 0, editor2.calls, "editor2 calls (one non-matching")
}

func TestNotInvertsMatch(t *testing.T) {
	editor1 := &recordingVisitor{}
	Check(Not(&markingVisitor{}), editor1).Visit(newSourceFile(), nil)
	assert.Equal(t, 0, editor1.calls, "not(matching): editor calls")

	editor2 := &recordingVisitor{}
	Check(Not(&recordingVisitor{}), editor2).Visit(newSourceFile(), nil)
	assert.Equal(t, 1, editor2.calls, "not(non-matching): editor calls")
}

func TestBareRecipeRefShortCircuitsToMatch(t *testing.T) {
	// A bare RecipeRef without a LocalVisitor short-circuits to "matches"
	// in-process so the wrapped editor still runs. The host evaluates the
	// gate for real once the response goes over the wire.
	editor := &recordingVisitor{}
	bare := &RecipeRef{
		RecipeName: "org.openrewrite.java.search.HasMethod",
		Options:    map[string]any{"methodPattern": "*..* nope(..)"},
	}
	Check(bare, editor).Visit(newSourceFile(), nil)
	assert.Equal(t, 1, editor.calls, "editor calls (bare RecipeRef")
}

func TestRecipeRefWithLocalVisitorEvaluatesForReal(t *testing.T) {
	// Helpers like UsesMethod populate a native LocalVisitor so unit
	// tests without an active RPC connection still see real filtering.
	// An empty CompilationUnit has no method invocations, so the gate
	// fails and the editor is skipped.
	editor := &recordingVisitor{}
	Check(UsesMethod("*..* tostring(..)"), editor).Visit(newSourceFile(), nil)
	assert.Equal(t, 0, editor.calls, "editor calls (RecipeRef with LocalVisitor")
}

func TestHelpersPopulateLocalVisitor(t *testing.T) {
	// Spot-check that helpers bundle a TreeVisitor for offline eval.
	assert.NotNil(t, HasSourcePath("**/*.go").LocalVisitor, "HasSourcePath did not populate LocalVisitor")
	assert.NotNil(t, UsesMethod("*..* a(..)").LocalVisitor, "UsesMethod did not populate LocalVisitor")
	assert.NotNil(t, UsesType("foo.Bar").LocalVisitor, "UsesType did not populate LocalVisitor")
	assert.NotNil(t, FindMethods("*..* a(..)").LocalVisitor, "FindMethods did not populate LocalVisitor")
	assert.NotNil(t, FindTypes("foo.Bar").LocalVisitor, "FindTypes did not populate LocalVisitor")
}

func TestHasSourcePathMatchesCompilationUnit(t *testing.T) {
	// given a real source file with a path
	editor := &recordingVisitor{}
	cu := &golang.CompilationUnit{SourcePath: "pkg/foo/bar.go"}

	// when the glob matches the path
	Check(HasSourcePath("**/*.go"), editor).Visit(cu, nil)

	// then the editor runs
	assert.Equal(t, 1, editor.calls, "editor calls (matching path")

	// and when the glob does not match
	editor2 := &recordingVisitor{}
	Check(HasSourcePath("**/*.java"), editor2).Visit(cu, nil)
	assert.Equal(t, 0, editor2.calls, "editor calls (non-matching path")
}

func TestHasSourcePathMatchesGoMod(t *testing.T) {
	// given a go.mod root (a non-J SourceFile)
	editor := &recordingVisitor{}
	mod := &golang.GoMod{SourcePath: "go.mod"}

	// when the glob matches
	Check(HasSourcePath("**/go.mod"), editor).Visit(mod, nil)

	// then the editor runs even though GoMod is not a java.SourceFile
	assert.Equal(t, 1, editor.calls, "editor calls (go.mod")

	// and a non-matching glob must actually filter the GoMod out, rather
	// than the gate being bypassed because GoMod is not a java.SourceFile.
	editor2 := &recordingVisitor{}
	Check(HasSourcePath("**/*.go"), editor2).Visit(mod, nil)
	assert.Equal(t, 0, editor2.calls, "editor calls (go.mod, non-matching")
}

func TestUsesMethodMatchesInvocationInTree(t *testing.T) {
	// given a compilation unit containing fmt.Println(...)
	mi := &java.MethodInvocation{
		Select: &java.RightPadded[java.Expression]{Element: &java.Identifier{Name: "fmt", Type: &java.JavaTypeClass{FullyQualifiedName: "fmt"}}},
		Name:   &java.Identifier{Name: "Println"},
	}
	cu := &golang.CompilationUnit{
		Statements: []java.RightPadded[java.Statement]{{Element: mi}},
	}

	// when gating on the matching method pattern
	editor := &recordingVisitor{}
	Check(UsesMethod("fmt Println(..)"), editor).Visit(cu, nil)

	// then the editor runs
	assert.Equal(t, 1, editor.calls, "editor calls (matching method")

	// and a non-matching pattern skips the editor
	editor2 := &recordingVisitor{}
	Check(UsesMethod("fmt Printf(..)"), editor2).Visit(cu, nil)
	assert.Equal(t, 0, editor2.calls, "editor calls (non-matching method")
}

func TestUsesTypeMatchesAttributionInTree(t *testing.T) {
	// given a node carrying type attribution for "tarfile"
	typed := &java.Identifier{
		Name: "tr",
		Type: &java.JavaTypeClass{FullyQualifiedName: "tarfile"},
	}
	mi := &java.MethodInvocation{Name: typed}
	cu := &golang.CompilationUnit{
		Statements: []java.RightPadded[java.Statement]{{Element: mi}},
	}

	// when gating on the matching type
	editor := &recordingVisitor{}
	Check(UsesType("tarfile"), editor).Visit(cu, nil)

	// then the editor runs
	assert.Equal(t, 1, editor.calls, "editor calls (matching type")

	// and an unrelated type skips the editor
	editor2 := &recordingVisitor{}
	Check(UsesType("zipfile"), editor2).Visit(cu, nil)
	assert.Equal(t, 0, editor2.calls, "editor calls (non-matching type")
}

func TestOrRequiresAtLeastTwoOperands(t *testing.T) {
	defer func() {
		if r := recover(); r == nil {
			t.Errorf("expected panic on single-operand Or")
		}
	}()
	Or(&recordingVisitor{})
}

func TestAndRequiresAtLeastTwoOperands(t *testing.T) {
	defer func() {
		if r := recover(); r == nil {
			t.Errorf("expected panic on single-operand And")
		}
	}()
	And(&recordingVisitor{})
}
